storage: drain eagerly after lease transfers

tbg opened the pull request:
The decision to introduce a 5s sleep in the draining code was originally
made due to its simplicity[1]. However, lots of tests drain and today all
of them sleep for multiple seconds, which is generally undesirable.

For example, `TestRangeTransferLeaseExpirationBased` drains twice and thus
takes at least ten seconds before this PR.

This PR makes this code smarter; it will try for at most 5s to transfer
leases away but calls it a day once it has only followers (or
singlep-replica ranges). This gives some nice extra coverage of things
working as expected, measured by the lease transfer tests now taking
~250ms.

The other option would have been to manually override this sleep timeout in
tests, but we would likely want that done selectively in tests that need
it, which makes new or changed tests susceptible to picking the delay up
again. Plus, it's generally a good thing if tests and production code
behave in the same way.

Fixes #24666.

[1]:
https://github.com/cockroachdb/cockroach/issues/22573#issuecomment-366033405

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25066)
<!-- Reviewable:end -->

bdarnell commented:
Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/store.go, line 1014 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LAy9Zie2FA7hoKc1dOD:-LAy9Zie2FA7hoKc1dOE:buuz8uy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49dcaed48d1cfedf058ceeebfe00127fd3b68d/pkg/storage/store.go#L1014)):*
> ```Go
> 					r.mu.Lock()
> 					r.mu.draining = true
> 					status := *r.raftStatusRLocked()
> ```

`r.raftStatusRLocked()` can return nil. 

---

*[pkg/storage/store.go, line 1018 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LAyDZa_28XgSI1s1GrA:-LAyDZa_28XgSI1s1GrB:boocarh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49dcaed48d1cfedf058ceeebfe00127fd3b68d/pkg/storage/store.go#L1018)):*
> ```Go
> 					// indicates follower status, or if it indicates a single-replica Raft
> 					// group (in which case we can't transfer the lease anywhere).
> 					isFollower := (status.RaftState == raft.StateFollower && status.Lead != 0) || len(status.Progress) <= 1
> ```

Adding all these extra conditions to a variable named `isFollower` is confusing. I think I'd invert it and call it `needsRaftTransfer`. The comments should also explain *why* we're using this set of conditions ("If there is another replica that could become leader, wait for the raft election to complete. As long as the lease and raft leadership are not colocated, the replica will continually try to transfer raft leadership.")

Can we put all the RaftStatus stuff into the `else` of `if ownsValidLease` instead of adding something here that we may have to undo later?

---

*[pkg/storage/store.go, line 1067 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LAyComvE7I6eEmUOdh_:-LAyComvE7I6eEmUOdha:bov964q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49dcaed48d1cfedf058ceeebfe00127fd3b68d/pkg/storage/store.go#L1067)):*
> ```Go
> 							// follower (though that means that nothing will be checked in
> 							// the next iteration).
> 							atomic.AddInt32(&transferCount, 1)
> ```

This double-counts with the `if !isFollower` path above. That doesn't matter since we just use this count to see if it's non-zero, but maybe that means a counter isn't really the right model anyway. Maybe a `stillWorking` bool would be better. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25066)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/store.go, line 1014 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LAy9Zie2FA7hoKc1dOD:-LB1njukBXSpa94igfQK:b-lhm5nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49dcaed48d1cfedf058ceeebfe00127fd3b68d/pkg/storage/store.go#L1014)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`r.raftStatusRLocked()` can return nil. 
</blockquote></details>

 Good catch, done.

---

*[pkg/storage/store.go, line 1018 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LAyDZa_28XgSI1s1GrA:-LB1nhqPBBA9LwDADdM0:b-8y5ou6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49dcaed48d1cfedf058ceeebfe00127fd3b68d/pkg/storage/store.go#L1018)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Adding all these extra conditions to a variable named `isFollower` is confusing. I think I'd invert it and call it `needsRaftTransfer`. The comments should also explain *why* we're using this set of conditions ("If there is another replica that could become leader, wait for the raft election to complete. As long as the lease and raft leadership are not colocated, the replica will continually try to transfer raft leadership.")

Can we put all the RaftStatus stuff into the `else` of `if ownsValidLease` instead of adding something here that we may have to undo later?
</blockquote></details>

Done, how's this?

---

*[pkg/storage/store.go, line 1067 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LAyComvE7I6eEmUOdh_:-LB1noNRDYLtT4REIOHv:bj4fvy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd49dcaed48d1cfedf058ceeebfe00127fd3b68d/pkg/storage/store.go#L1067)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This double-counts with the `if !isFollower` path above. That doesn't matter since we just use this count to see if it's non-zero, but maybe that means a counter isn't really the right model anyway. Maybe a `stillWorking` bool would be better. 
</blockquote></details>

How's this?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25066)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 423 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LB2IPyP3hdp18vTenHE:-LB2IPyQAG5qFf0hN428:b-n0uimb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5014867f54e90cd59262ba6e1a17c5afc4643543/pkg/storage/replica_proposal.go#L423)):*
> ```Go
> // is behind on applying the log.
> //
> // TODO(tschottdorf): there's also maybeTransferRaftLeadership. Only one should exist.
> ```

s/ship//

---

*[pkg/storage/store.go, line 1061 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LB2KB3xDi40rRQXxV4w:-LB2KB3xDi40rRQXxV4x:bc0uu6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/5014867f54e90cd59262ba6e1a17c5afc4643543/pkg/storage/store.go#L1061)):*
> ```Go
> 							}
> 						}
> 						transferred, err := s.replicateQueue.transferLease(
> ```

s/transferred/leaseTransferred/

---

*[pkg/storage/store.go, line 1081 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LB2K5XT1zPFI3U6MMHq:-LB2K5XT1zPFI3U6MMHr:b-wnta0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/5014867f54e90cd59262ba6e1a17c5afc4643543/pkg/storage/store.go#L1081)):*
> ```Go
> 					}
> 
> 					if needsRaftTransfer {
> ```

I think we only want to do this if `!transferred`. If we transfer lease and leadership at the same time it will just take longer. (and normally transferring a lease triggers a leadership transfer attempt when it finishes)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-:-LB2KWc47AU2NVl-KY52:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/replica_proposal.go, line 423 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LB2IPyP3hdp18vTenHE:-LBdynna7HKlVp0fEtJK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5014867f54e90cd59262ba6e1a17c5afc4643543/pkg/storage/replica_proposal.go#L423)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/ship//
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1061 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LB2KB3xDi40rRQXxV4w:-LBdytrJ-W-6DRMQAxQC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5014867f54e90cd59262ba6e1a17c5afc4643543/pkg/storage/store.go#L1061)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/transferred/leaseTransferred/
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1081 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/25066#-LB2K5XT1zPFI3U6MMHq:-LBdzJAq73AHGXLApv5Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5014867f54e90cd59262ba6e1a17c5afc4643543/pkg/storage/store.go#L1081)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think we only want to do this if `!transferred`. If we transfer lease and leadership at the same time it will just take longer. (and normally transferring a lease triggers a leadership transfer attempt when it finishes)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25066)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: drain eagerly after lease transfers

The decision to introduce a 5s sleep in the draining code was originally
made [due to its simplicity1]. However, lots of tests drain and today
all of them sleep for multiple seconds, which is generally undesirable.

For example, `TestRangeTransferLeaseExpirationBased` drains twice and
thus takes at least ten seconds before this PR.

This PR makes this code smarter; it will try for at most 5s to transfer
leases away but calls it a day once it has only followers (or
singlep-replica ranges). This gives some nice extra coverage of things
working as expected, measured by the lease transfer tests now taking
~250ms.

The other option would have been to manually override this sleep timeout
in tests, but we would likely want that done selectively in tests that
need it, which makes new or changed tests susceptible to picking the
delay up again. Plus, it's generally a good thing if tests and
production code behave in the same way.

Fixes #24666.

[due to its simplicity]: https://github.com/cockroachdb/cockroach/issues/22573#issuecomment-366033405

Release note: None

tschottdorf commented:
CI flaked on  TestDockerCLI/test_log_config_msg.tcl. These tests are not legible at all but I think this is the problem?

```
.180504 11:14:10.925984999 EXPECT TEST: START TEST: Check that the cluster ID is reported at the start of new log files.
child process exited abnormally
    while executing
"system "grep '\\\[config\\\] clusterID:' logs/db/logs/cockroach.log""
    (file "/go/src/github.com/cockroachdb/cockroach/cli/interactive_tests/test_log_config_msg.tcl" line 17)
non-zero exit code: 1
    	cli_test.go:92: non-zero exit code: 1

```

This doesn't seem to be related to this change, so I'm going to run again and merge if it's successful.

tschottdorf commented:
Filed #25334.

bors r=bdarnell
